File: day-01/rust/src/main.rs
fn part1(input: &str) -> u32 {
    let mut total = 0;
    
    input.split('\n').for_each(|line| {
        let mut last = None;
        let mut first = None;

        line.chars().for_each(|c| {
            if c.is_digit(10) {
                let digit = c.to_digit(10).unwrap();

                if last.is_none() {
                    first = Some(digit);
                }

                last = Some(digit);
            }
        });

        total += first.unwrap() * 10 + last.unwrap();
    });

    return total;
}

const NUMBERS: [&str; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"
];

fn part2(input: &str) -> u32 {
    let mut total = 0;
    
    input.split('\n').for_each(|line| {
        let mut line = line.to_string();

        let mut last = None;
        let mut first = None;

        while line.len() > 0 {
            for (i, number) in NUMBERS.iter().enumerate() {
                if line.starts_with(number) {
                    if last.is_none() {
                        first = Some(i as u32 + 1);
                    }

                    last = Some(i as u32 + 1);
                    break;
                }

                let next = line.chars().next().unwrap();
                if next.is_digit(10) {
                    let digit = next.to_digit(10).unwrap();

                    if last.is_none() {
                        first = Some(digit);
                    }

                    last = Some(digit);
                    break;
                }
            }

            line.remove(0);
        }

        total += first.unwrap() * 10 + last.unwrap();
    });

    return total;
}

fn main() {
    let input = include_str!("../../input.txt");

    println!("Part 1: {}", part1(input));
    println!("Part 2: {}", part2(input));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() {
        assert_eq!(part1("1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet"), 142);
    }

    #[test]
    fn test_part2() {
        assert_eq!(part2("two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen"), 281);
    }
}

File: day-01/c/src/main.c
#include <stdio.h>
#include <string.h>
#include <assert.h>
#include <stdlib.h>
#include <stdbool.h>

#include "../../../lib/c/get_input.h"

char *NUMEBERS[9] = {
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine"
};

int part1(const char *input) {
    int total = 0;

    while (*input) {
        int first = 0;
        int last = 0;
        char c;

        while ((c = *input++)) {
            c -= '0';

            if (c >= 1 && c <= 9) {
                if (!first) {
                    first = c;
                }

                last = c;
            }

            if (*input == '\n') {
                input++;
                break;
            }
        }

        total += first * 10 + last;
    }

    return total;
}

bool starts_with(const char *str, const char *prefix) {
    while (*prefix) {
        if (*str++ != *prefix++) {
            return false;
        }
    }

    return true;
}

int part2(const char *input) {
        int total = 0;

    while (*input) {
        int first = 0;
        int last = 0;
        char c;

        while ((c = *input)) {
            c -= '0';

            if (c >= 1 && c <= 9) {
                if (!first) {
                    first = c;
                }

                last = c;
            }

            for (int i = 0; i < 9; i++) {
                if (starts_with(input, NUMEBERS[i])) {
                    if (!first) {
                        first = i + 1;
                    }

                    last = i + 1;
                }
            }

            input++;
            if (*input == '\n') {
                input++;
                break;
            }
        }

        total += first * 10 + last;
    }

    return total;
}

void test();
int main(int argc, char **argv) {
    if (argc > 1 && strcmp(argv[1], "--test") == 0) {
        test();
        return 0;
    }

    char* input = get_input(argc, argv);

    printf("Part 1: %d\n", part1(input));
    printf("Part 2: %d\n", part2(input));

    return 0;
}

void test() {
    printf("Testing part 1: ");
    fflush(stdout);

    assert(part1("1abc2\n\
pqr3stu8vwx\n\
a1b2c3d4e5f\n\
treb7uchet") == 142);

    printf("PASSED\nTesting part 2: ");
    fflush(stdout);

    assert(part2("two1nine\n\
eightwothree\n\
abcone2threexyz\n\
xtwone3four\n\
4nineeightseven2\n\
zoneight234\n\
7pqrstsixteen") == 281);


    printf("PASSED\n\nBoth tests passed!\n");
}